### PR TITLE
Fix rm for symlinks pointing to directory on windows (issue #11461)

### LINKED ISSUE #11461: rm: Cannot remove a symlink pointing to a directory on Windows
### Describe the bug

`rm` command results into an os error **on Windows** when trying to remove a symbolic link pointing to a directory (made with `mklink /D`) .

Perhaps this bug can be the same as #10899 , but I reproduced this in REPL.

### How to reproduce

(on Windows)

1. create a symlink pointing to a directory. ( `mklink /D link foo`; This operation requires administrator privilege.)
2. try remove the symlink. (`rm link`)

### Expected behavior

- I expect the symlink to be deleted.
- (Also) I expect the directory the link points to not to be deleted.

### Screenshots

image

message translation:
- `link <<===>> foo のシンボリック リンクが作成されました` -> `symbolic link created for link <<===>> foo `
- `Could not delete C:\Users\yt\work\tmp\link: アクセスが拒否されました。 (os error 5)`  ->  `Could not delete C:\Users\yt\work\tmp\link: Access denied. (os error 5)`

### Configuration

| key                | value                                                |
| ------------------ | ---------------------------------------------------- |
| version            | 0.88.1                                               |
| branch             |                                                      |
| commit_hash        | fd56768fdcc1acbebcb75745600fc7419f72833d             |
| build_os           | windows-x86_64                                       |
| build_target       | x86_64-pc-windows-msvc                               |
| rust_version       | rustc 1.72.1 (d5c2e9c34 2023-09-13)                  |
| rust_channel       | 1.72.1-x86_64-pc-windows-msvc                        |
| cargo_version      | cargo 1.72.1 (103a7ff2e 2023-08-15)                  |
| build_time         | 2023-12-14 17:47:47 +00:00                           |
| build_rust_channel | release                                              |
| allocator          | mimalloc                                             |
| features           | dataframe, default, extra, sqlite, trash, which, zip |
| installed_plugins  |                                                      |

### Additional context

According to [Microsoft's document of CreateSymbolicLinkA function](https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createsymboliclinka):

> To remove a symbolic link, delete the file (using [DeleteFile](https://learn.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-deletefilea) or similar APIs) or remove the directory (using [RemoveDirectory](https://learn.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-removedirectorya) or similar APIs) depending on what type of symbolic link is used.

,but currently `rm` tries to [`std::fs::remove_file`](https://doc.rust-lang.org/std/fs/fn.remove_file.html) (same as `DeleteFile` on Windows) regardless of what the symlink pointing to. I think `rm` should use [`std::fs::remove_dir`](https://doc.rust-lang.org/std/fs/fn.remove_dir.html) (same as `RemoveDirectory` on Windows) instead when the symlink pointing to a directory.

https://github.com/nushell/nushell/blob/f5973801129a13d834af9f0e738d182e49cc2e3b/crates/nu-command/src/filesystem/rm.rs#L389C24-L389C24

### PR DESCRIPTION
**Pull request by yukitomoda**
- this PR closes #11461

# Description

Using `std::fs::remove_dir` instead of `std::fs::remove_file` when try remove symlinks pointing to a directory on Windows.

# User-Facing Changes

none

# Tests + Formatting

- [x] `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- [x] `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used` to check that you're using the standard code style
- [x] `cargo test --workspace` to check that all tests pass (on Windows make sure to [enable developer mode](https://learn.microsoft.com/en-us/windows/apps/get-started/developer-mode-features-and-debugging))
  - I got 2 test fails on my Windows devenv; these fails in main branch too
     - `commands::complete::basic` : passed on Ubuntu, failed on Windows (a bug?)
     - `commands::cp::copy_file_with_read_permission`: failed on Windows with Japanese environment (This test refers error message, so that fails on environments using a language except for english.)
- [x] `cargo run -- -c "use std testing; testing run-tests --path crates/nu-std"` to run the tests for the standard library

# After Submitting

This fix has no changes to user-facing interface.

### CODE REVIEW

**Comment by yukitomoda** [comment]
As above, I got 2 test failure on my Windows (Japanese), but these passed on my ubuntu.

## `commands::complete::basic`

https://github.com/nushell/nushell/blob/f5973801129a13d834af9f0e738d182e49cc2e3b/crates/nu-command/tests/commands/complete.rs#L4C1-L10

I think this is caused by `\n` or `\r\n`, but have no idea why the first one passed on CI...

```nu
~\work\tmp
❯ (^echo a | complete) == {stdout: "a\n", exit_code: 0}
false

~\work\tmp
❯ (^echo a | complete) == {stdout: "a\r\n", exit_code: 0}
true
```

## `commands::cp::copy_file_with_read_permission`

https://github.com/yukitomoda/nushell/blob/de5ad5de19affee3986661da6dd888a2e12a813f/crates/nu-command/tests/commands/cp.rs#L576-L582

I think this is caused by referencing error messages in this test.

File: crates/nu-command/src/filesystem/rm.rs [hunk 1]
```diff
@@ -381,11 +381,24 @@ fn rm(
...
-                    } else if metadata.is_file()
-                        || is_socket
-                        || is_fifo
-                        || metadata.file_type().is_symlink()
-                    {
+                    } else if metadata.is_symlink() {
+                        // In Windows, symlink pointing to a directory can be removed using
+                        // std::fs::remove_dir instead of std::fs::remove_file.
+                        #[cfg(windows)]
+                        match f.metadata() {
```
Comment (WindSoilder) [inline crates/nu-command/src/filesystem/rm.rs:388]: What about using `and_then`?
```rs
f.metadata().and_then(|metadata|
    if metadata.is_dir() { std::fs::remove_dir(&f) }
    else { std::fs::remove_file(&f) }
}?
```

**Comment by WindSoilder** [comment]
> As above, I got 2 test failure on my Windows (Japanese), but these passed on my ubuntu.

I remember it's because windows CI have git bash installed

File: crates/nu-command/src/filesystem/rm.rs (hunk 1 above)
Comment (yukitomoda) [reply crates/nu-command/src/filesystem/rm.rs]: > What about using and_then?

Indeed, it's a better way.
I'm going to refactor it.

**Comment by yukitomoda** [comment]
> > As above, I got 2 test failure on my Windows (Japanese), but these passed on my ubuntu.
> 
> I remember it's because windows CI have git bash installed

Oh, it makes sense.
I tried `cargo test --workspace` in git bash and `commands::complete::basic` passed.

Then, should we change the tests to be independent on platforms? Or should we note in CONTRIBUTING.md that `cargo test --workspace` should be runned in git bash on Windows?

**Comment by fdncred** [comment]
We should not rely on git bash being installed.

**Review by WindSoilder — APPROVED** [review]
Thanks!  LGTM!

For the failing test on windows, I think we can cover it in separate pr :-D
Thanks for your contribution!

**Pull request merged**